### DISCUSSION TITLE
Run homebrew CI locally

### DISCUSSION THREAD

**Discussion by hugoclrd** [Homebrew/discussions#4517] [Tap maintenance and Homebrew development]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.18-85-g16b2a14
ORIGIN: https://github.com/Homebrew/brew
HEAD: 16b2a14c9408140b04828462b660d30056421fdb
Last commit: 22 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: faa43e4d84466c71b4c6a2ba2e82f5f8eff80b57
Core tap last commit: 5 hours ago
Core tap branch: master
Core tap JSON: 19 May 11:12 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.40.1 => /opt/homebrew/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
/
```

### Description of issue

Hello,

I'm trying to understand what is happening here: https://github.com/Homebrew/homebrew-core/pull/130619 and here https://github.com/Homebrew/homebrew-core/pull/131127. The CI fails and it's hard to understand what is actually happening.

I'd like to reproduce the issue locally, is it possible to run Homebrew's GitHub actions locally?
I've seen this tool call  `ask` but I haven't been able to use it so far.

Thanks 

Note: if you have another ideas on how to debug the build, I'm happy to hear as well 🙏

**Comment by dawidd6** [comment]
`brew test-bot $FORMULA` should get you going easily.

**Reply by hugoclrd** [reply]
It seems like it could be useful indeed. How can I build a specific tag or branch of the formula?

**Reply by SMillerDev** [reply]
That's not supported in brew CI.

**Reply by carlocab** [reply]
Just check out the branch you want at `"$(brew --repository homebrew/core)"`.

**Reply by hugoclrd** [reply]
Yes I updated the local clarinet.rb file with the version I want to test. 

And I'm trying to run
```
HOMEBREW_NO_INSTALL_FROM_API=1 && brew test-bot clarinet --verbose
```

I tried with the flag `--build-from-source` but it leads to the error "Not building a bottle for clarinet because it has unbottled dependencie"

Anyway, I still haven't been able to reproduce the error happening in homebrew CI and I soon out of ideas to debug it

**Reply by carlocab** [reply]
Hmmm. Can you show the output of `brew linkage clarinet` after a successful build locally?

**Reply by hugoclrd** [reply]
Sure:

```
brew linkage clarinet

System libraries:
  /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
  /System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices
  /System/Library/Frameworks/Security.framework/Versions/A/Security
  /usr/lib/libSystem.B.dylib
  /usr/lib/libiconv.2.dylib
```

**Reply by hugoclrd** [reply]
Thanks for restarting the CI (I think someone did)

**Reply by gromgit** [reply]
_Sidenote_: I'm sure you already know this, but many others probably don't. This command:-
> ```
> HOMEBREW_NO_INSTALL_FROM_API=1 && brew test-bot clarinet --verbose
> ```

sets the _shell variable_ `HOMEBREW_NO_INSTALL_FROM_API`, and _does not export_ it to your shell environment, so `brew test-bot` *cannot see it*.

You should instead run either:-
```
HOMEBREW_NO_INSTALL_FROM_API=1 brew test-bot clarinet --verbose
```
or:-
```
export HOMEBREW_NO_INSTALL_FROM_API=1; brew test-bot clarinet --verbose
```
both of which set the _environment variable_ `HOMEBREW_NO_INSTALL_FROM_API` that `brew test-bot` _does_ see. The only difference is that the first one sets the variable _only_ for `brew test-bot`, while the second sets it for all subsequent commands as well.

**Comment by hugoclrd** [comment]
Update: We're now sure that the issue come from the Deno version we are embedding in Clarinet. Unfortunately, it will take us a bit time to fix it (upgrade Deno or get rid of it).
As a short term solution, it has been suggested (https://github.com/Homebrew/homebrew-core/pull/131430#issuecomment-1567621542) to only build for macos 11-arm64 (and not for 12 and 13). This solution sound acceptable to me.

But I need some help/guidance to achieve it

**Reply by SMillerDev** [reply]
Homebrew formulae shouldn't embed other software available in Homebrew: https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae

**Reply by hugoclrd** [reply]
Thanks for your answer @SMillerDev 

Well I agree that the way we vendor Deno isn't the best design, and we plan on removing it in Q3, it'll take some time and efforts.
But we've been shipping it like it for months without issue (see v1.5.4). Until homebrew's CI started to break for unknow reasons (so far). 

So we have new features and bug fixes to ship that are currently blocked because we can't pass the CI right now. So I'm looking for solutions in the meantime. And the one suggested (skip build for macos 12 and 13) sounds good to me

**Comment by hugoclrd** [comment]
Hello,

We are still having issue to release, although there have been a solution suggested but no follow-up. And I don't know how to proceed.
I really need some help.
Should I open an issue? Is it possible to have a chat with someone from the team?
Thank you.

Note: the issue doesn't really match with the original topic of this discussion but I'm stuck and don't what to do to get help from the team.

cc @Bo98 @SMillerDev @carlocab

**Reply by MikeMcQuaid** [reply]
Don't open an issue, people will reply here.

**Reply by hugoclrd** [reply]
OK thanks, I wasn't sure this was the best place.

Honestly I feel like the actual issue is being ignored

**Reply by MikeMcQuaid** [reply]
It's not being ignored, it's just this is an open source project run by volunteers and several people have tried to help as much as they can but we have many thousands of packages and far fewer maintainers and need to prioritise.

**Reply by hugoclrd** [reply]
Yes I know, and I'm grateful the help I had so far. 
But we are stagnating. 
I'd like to understand why something that used to work doesn't work anymore (and then maybe, work on solution on our side or with the homebrew team)
